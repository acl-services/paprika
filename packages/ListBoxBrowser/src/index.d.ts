export default ListBoxBrowser;

declare function ListBoxBrowser(props: ListBoxBrowserProps): JSX.Element;
interface ListBoxBrowserProps {
  /** An array of javascript objects holding the data structure for the ListBoxBrowser. The object shape must have at least a string label property and an array options property in one of the objects. Also can hold any other kind of data for your own use. */
  data: any;
  /** Indicates if the user can select multiple options */
  isMulti?: boolean;
  /** Set the height for the ListBoxBrowser */
  height?: number;
  /** Callback function receiving an array of selected options by the component */
  onChange?: (...args: any[]) => any;
  /** Allows the user to select the parent options */
  isParentSelectable?: boolean;
  /** Content title for the left column */
  rootTitle?: React.ReactNode;
  /** Content title for the right column */
  browserTitle?: React.ReactNode;
  /** You can pass <ListBoxBrowser.OptionSelected /> as a children */
  children?: React.ReactNode;
  /** Indicates if the right column should display a breadcrumb */
  hasBreadcrumb?: boolean;
  /** Set a border red color around the component indicating that has an error */
  hasError?: boolean;
  /** When declaring the array options empty, this will be executed to retrieve the data, useful if you want to do a lazy load. */
  onFetch?: (...args: any[]) => any;
  /** A function that sets an option selected returning true or false you can use to compare your data structure and decide if the option is initially selected or not. */
  defaultSelectedOptions?: (...args: any[]) => any;
  /** A function that sets the initial view for the right columns (Browser) of the ListBoxBrowser the option selected to be the initial view should have options to be valid, by default the ListBoxBrowser picked the first option which has options to be the initial value. */
  defaultSelectedView?: (...args: any[]) => any;
  /** In the case you want to use the ListBoxBrowser with one column you can hide the root column */
  hasLeftColumn?: boolean;
}
declare function CustomListBox(props: CustomListBoxProps): JSX.Element;
interface CustomListBoxProps {
  hasOnUp?: boolean;

  id?: string | number;

  isLoading?: boolean;

  onChange: (...args: any[]) => any;

  onClickNavigate: (...args: any[]) => any;

  onUp?: (...args: any[]) => any;

  options: object[];
}
declare function Title(props: TitleProps): JSX.Element;
interface TitleProps {
  rootTitle: React.ReactNode;

  browserTitle: React.ReactNode;

  data?: object[];

  browserKey: string | number;

  onClickBreadcrumb: (...args: any[]) => any;

  hasLeftColumn: boolean;
}
