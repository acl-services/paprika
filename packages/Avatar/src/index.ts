export { default } from "./Avatar";
export * from "./helpers";
