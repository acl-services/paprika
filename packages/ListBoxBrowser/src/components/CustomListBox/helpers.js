export function isSelected({ $$key, selectedOptions, browserKey, isRootListBox }) {
  let key = browserKey;
  if (isRootListBox) {
    key = "root";
  }

  if (key in selectedOptions) {
    return selectedOptions[key].some(option => option.$$key === $$key);
  }

  return false;
}

export function isSelectable({ hasOptions, isParentSelectable }) {
  if (isParentSelectable !== null) {
    return !isParentSelectable;
  }

  return hasOptions;
}
