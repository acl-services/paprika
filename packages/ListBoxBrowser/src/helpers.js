import React from "react";

/** helpers */
const hasOptions = options => typeof options !== "undefined" && Array.isArray(options);

export function isRoot(key) {
  if (key === null) return false;

  return key.includes("root");
}

function getOptions(indexes, list) {
  if (Array.isArray(indexes)) {
    return indexes.map(index => list[index].value);
  }

  const options = [list[indexes].value].filter(chunk => chunk);
  return options.length ? options : null;
}

function isSelectable({ hasOptions, isParentSelectable }) {
  if (isParentSelectable !== null) {
    return !isParentSelectable;
  }

  return hasOptions;
}

/** DATA STRUCTURE */

export function getData({
  data,
  path = "root",
  refDefaultSelectedOptions,
  defaultSelectedOptions,
  defaultSelectedView,
  refDefaultSelectedView,
}) {
  return data.map(({ label = null, options, ...moreAttributes }, index) => {
    if (!label) {
      throw new Error("A 'label' attribute is required for each option object, can't process data.");
    }

    let newPath = isRoot(path) ? `${index}` : `${path}/${index}`;
    const recursiveParameters = {
      data: options,
      defaultSelectedOptions,
      defaultSelectedView,
      path: newPath,
      refDefaultSelectedOptions,
      refDefaultSelectedView,
    };
    const _hasOptions = hasOptions(options);

    const option = {
      parent: path,
      hasOptions: _hasOptions,
      $$key: newPath,
      attributes: {
        label,
        options: _hasOptions ? getData(recursiveParameters) : null,
        ...moreAttributes,
      },
    };

    if (defaultSelectedOptions(option.attributes)) {
      /* eslint-disable no-param-reassign */
      if (isRoot(option.parent)) {
        newPath = "root";
      }

      if (option.parent in refDefaultSelectedOptions.current) {
        refDefaultSelectedOptions.current = {
          ...refDefaultSelectedOptions.current,
          [option.parent]: [...refDefaultSelectedOptions.current[option.parent], option],
        };
      } else {
        refDefaultSelectedOptions.current = { ...refDefaultSelectedOptions.current, [option.parent]: [option] };
      }
    }

    // checking if its initial view
    if (_hasOptions && defaultSelectedView && defaultSelectedView(option.attributes)) {
      refDefaultSelectedView.current = newPath;
    }

    return option;
  });
}

export function getOptionByKey(data, path) {
  let node = null;
  function runner(data, path) {
    return data.map(option => {
      if (!option.$$key.includes("/") && typeof data[path] !== "undefined" && option.$$key === path) {
        node = option;
        throw new Error("option found");
      }

      if (option.$$key === path) {
        node = option;
        throw new Error("option found");
      }

      if (hasOptions(option.attributes.options)) {
        return runner(option.attributes.options, path);
      }

      return [];
    });
  }

  try {
    runner(data, path);
    return [];
  } catch (e) {
    return node;
  }
}

export function getBreadcrumb({ data, option }) {
  let activeOption = option;
  let key = null;
  const breadcrumb = [];
  while (!isRoot(key)) {
    if (!isRoot(activeOption.parent)) {
      const parent = getOptionByKey(data, activeOption.parent);
      breadcrumb.push(parent);
      activeOption = parent;
    }
    key = activeOption.parent;
  }

  return breadcrumb.reverse();
}

export function onChange({ source, indexes, list, isParentSelectable, setSelectedOptions, browserKey, isMulti }) {
  if (!isMulti) {
    setSelectedOptions(selectedOptions => {
      const key = isRoot(source) ? "root" : browserKey;
      if (Object.keys(selectedOptions).length) {
        const option = getOptions(indexes, list);
        if (option === null) return {};

        const isSameOption = Object.keys(selectedOptions).some(
          key => selectedOptions[key][0].$$key === option[0].$$key
        );

        if (isSameOption) {
          return {};
        }
      }

      const options = getOptions(indexes, list);
      if (options === null) return {};
      return { [key]: options };
    });

    return;
  }

  if (isRoot(source)) {
    const options = getOptions(indexes, list)
      .map(option => {
        if (isSelectable({ hasOptions: option.hasOptions, isParentSelectable })) {
          return null;
        }

        return option;
      })
      .filter(chunk => chunk);

    if (options.length) {
      setSelectedOptions(selectedOptions => ({ ...selectedOptions, root: [...options] }));
      return;
    }

    setSelectedOptions(selectedOptions => ({ ...selectedOptions, root: [] }));
  }

  setSelectedOptions(selectedOptions => ({ ...selectedOptions, [browserKey]: getOptions(indexes, list) }));
}

export function focusListBoxBrowser($root, hasLeftColumn) {
  window.requestAnimationFrame(() => {
    const index = hasLeftColumn ? 1 : 0;
    $root.querySelectorAll('[data-pka-anchor="list-box-content-inline"]')[index].focus();
  });
}

export function focusListBoxRoot($root) {
  window.requestAnimationFrame(() => {
    $root.querySelectorAll('[data-pka-anchor="list-box-content-inline"]')[0].focus();
  });
}

export function getInitialView(localData) {
  let index = null;
  try {
    Object.keys(localData).some(key => {
      if (
        localData[key].hasOptions ||
        (localData[key].attributes.options && localData[key].attributes.options.length === 0)
      ) {
        index = key;
        throw new Error("index found");
      }

      return false;
    });

    return index;
  } catch (e) {
    return index;
  }
}

export function extractedExtendedProps(children) {
  const props = {};
  React.Children.map(children, child => {
    if (child === null) return null;

    if (child.type.displayName === "ListBoxBrowser.Root") {
      props.root = { ...child.props };
    }

    if (child.type.displayName === "ListBoxBrowser.Browser") {
      props.browser = { ...child.props };
    }
  });

  return { ...props, children };
}

export function getDataOptionByFn(data, fn) {
  let node = null;
  function runner(data, fn) {
    data.forEach(option => {
      if (fn(option)) {
        node = option;
        throw new Error("We have found the option");
      }

      if (hasOptions(option.options)) {
        runner(option.options, fn);
      }
    });
  }

  try {
    runner(data, fn);
    return node;
  } catch (e) {
    // will fire if an option is found
    return node;
  }
}
