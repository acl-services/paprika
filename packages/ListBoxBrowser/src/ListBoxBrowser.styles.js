import { css } from "styled-components";
import tokens from "@paprika/tokens/lib/tokens";
import { spacer } from "@paprika/stylers/lib/helpers";

export const flex = css`
  align-items: flex-start;
  display: flex;
  justify-content: center;
`;

export const container = css`
  border: 2px solid ${tokens.border.color};
  border-radius: 6px;
  min-height: 240px;
  position: relative;
  width: 100%;

  [data-ppk-anchor="listbox-content-inline"] {
    border: 0;
    flex-basis: 50%;
  }

  [data-ppk-anchor="listbox-box"] {
    border: 0;
  }

  [data-ppk-anchor="listbox-trigger"] {
    border: 0;
    display: none;
  }

  [data-ppk-anchor="listbox-trigger"] ~ [data-ppk-anchor="listbox-trigger"] {
    border-left: 1px solid ${tokens.border.color};
    display: block;
    height: 100%;
    left: 50%;
    position: absolute;
    top: 0px;
    width: 1px;
  }
`;

export const breakLine = css`
  border: 1px solid red;
  flex-basis: 100%;
  height: 0;
`;

export const title = css`
  background: ${tokens.backgroundColor.level0};
  border-bottom: 1px solid ${tokens.border.color};
  flex-grow: 1;
  padding: ${spacer(1)} ${spacer(1)} ${spacer(1)} 12px;
`;
