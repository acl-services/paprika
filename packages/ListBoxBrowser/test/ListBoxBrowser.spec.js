import data from "./fixtures";
import { getData, getOptionByKey } from "../src/helpers";

describe("ListBoxBrowser", () => {
  it("should return a data with proper keys", () => {
    const selectedOptions = { current: {} };
    const modifiedData = getData({ data, selectedOptions });

    expect(modifiedData).toMatchSnapshot();
    expect(selectedOptions).toMatchSnapshot();
  });

  it("should return option with $$key === 0", () => {
    const selectedOptions = { current: {} };
    const option = getOptionByKey(getData({ data, selectedOptions }), "0");
    expect(option).toMatchSnapshot();
  });

  it("should return option with $$key === 1", () => {
    const selectedOptions = { current: {} };
    const option = getOptionByKey(getData({ data, selectedOptions }), "1");
    expect(option).toMatchSnapshot();
  });

  it("should return option with $$key === 2/1", () => {
    const selectedOptions = { current: {} };
    const option = getOptionByKey(getData({ data, selectedOptions }), "2/1");
    expect(option).toMatchSnapshot();
  });

  it("should return option with $$key === 1/2/1", () => {
    const selectedOptions = { current: {} };
    const option = getOptionByKey(getData({ data, selectedOptions }), "1/2/1");
    expect(option).toMatchSnapshot();
  });
});
