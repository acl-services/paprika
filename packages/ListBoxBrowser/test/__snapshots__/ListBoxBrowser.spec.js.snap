// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBoxBrowser should return a data with proper keys 1`] = `
Array [
  Object {
    "$$key": "0",
    "attributes": Object {
      "defaultIsSelected": false,
      "label": "Marvel Universe",
      "options": Array [
        Object {
          "$$key": "0/0",
          "attributes": Object {
            "defaultIsSelected": false,
            "label": "Humans",
            "options": Array [
              Object {
                "$$key": "0/0/0",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Black Cat",
                  "options": Array [
                    Object {
                      "$$key": "0/0/0/0",
                      "attributes": Object {
                        "defaultIsSelected": false,
                        "label": "Akhenaten",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                    Object {
                      "$$key": "0/0/0/1",
                      "attributes": Object {
                        "defaultIsSelected": false,
                        "label": "Alistair Smythe",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                    Object {
                      "$$key": "0/0/0/2",
                      "attributes": Object {
                        "defaultIsSelected": false,
                        "label": "Angelo Fortunato",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                    Object {
                      "$$key": "0/0/0/3",
                      "attributes": Object {
                        "defaultIsSelected": false,
                        "label": "Arcade",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                    Object {
                      "$$key": "0/0/0/4",
                      "attributes": Object {
                        "defaultIsSelected": false,
                        "label": "Aria",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                    Object {
                      "$$key": "0/0/0/5",
                      "attributes": Object {
                        "defaultIsSelected": false,
                        "label": "Baal",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                  ],
                },
                "hasOptions": true,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/1",
                "attributes": Object {
                  "defaultIsSelected": true,
                  "label": "The Punisher",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/2",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Silver Sable",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/3",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Red Skull",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/4",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Vulture",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/5",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Falcon",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/6",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Kraven the hunter",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/7",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Bruce banner",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/8",
                "attributes": Object {
                  "defaultIsSelected": true,
                  "label": "Bucky",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/9",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Moon Knight",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/10",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "KingPin",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/11",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Doctor Octopus",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/12",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Elektra",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/13",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Black Widow",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/14",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Hawkeye",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/15",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Bullseye",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/16",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "TaskMaster",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/17",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Star-Lord",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/18",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "DareDevil",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/19",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Ikari",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/20",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "War Machine",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/21",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Nick Fury",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/22",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Hank Pym",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/23",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Captain America",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/24",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Iron Man",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
            ],
          },
          "hasOptions": true,
          "parent": "0",
        },
        Object {
          "$$key": "0/1",
          "attributes": Object {
            "defaultIsSelected": false,
            "label": "Mutants",
            "options": null,
          },
          "hasOptions": false,
          "parent": "0",
        },
        Object {
          "$$key": "0/2",
          "attributes": Object {
            "defaultIsSelected": false,
            "label": "Celestials",
            "options": null,
          },
          "hasOptions": false,
          "parent": "0",
        },
      ],
    },
    "hasOptions": true,
    "parent": "root",
  },
  Object {
    "$$key": "1",
    "attributes": Object {
      "defaultIsSelected": false,
      "label": "DC Universe",
      "options": Array [
        Object {
          "$$key": "1/0",
          "attributes": Object {
            "defaultIsSelected": false,
            "label": "Humans",
            "options": null,
          },
          "hasOptions": false,
          "parent": "1",
        },
        Object {
          "$$key": "1/1",
          "attributes": Object {
            "defaultIsSelected": false,
            "label": "Meta Humans",
            "options": null,
          },
          "hasOptions": false,
          "parent": "1",
        },
        Object {
          "$$key": "1/2",
          "attributes": Object {
            "defaultIsSelected": false,
            "label": "The Quintessence",
            "options": Array [
              Object {
                "$$key": "1/2/0",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Zeus",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "1/2",
              },
              Object {
                "$$key": "1/2/1",
                "attributes": Object {
                  "defaultIsSelected": false,
                  "label": "Ganthet",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "1/2",
              },
              Object {
                "$$key": "1/2/2",
                "attributes": Object {
                  "defaultIsSelected": true,
                  "label": "Shazam",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "1/2",
              },
              Object {
                "$$key": "1/2/3",
                "attributes": Object {
                  "defaultIsSelected": true,
                  "label": "Highfather",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "1/2",
              },
            ],
          },
          "hasOptions": true,
          "parent": "1",
        },
      ],
    },
    "hasOptions": true,
    "parent": "root",
  },
  Object {
    "$$key": "2",
    "attributes": Object {
      "defaultIsSelected": false,
      "label": "Starcraft Universe",
      "options": Array [
        Object {
          "$$key": "2/0",
          "attributes": Object {
            "defaultIsSelected": false,
            "label": "Terrans",
            "options": null,
          },
          "hasOptions": false,
          "parent": "2",
        },
        Object {
          "$$key": "2/1",
          "attributes": Object {
            "defaultIsSelected": false,
            "label": "Zergs",
            "options": null,
          },
          "hasOptions": false,
          "parent": "2",
        },
        Object {
          "$$key": "2/2",
          "attributes": Object {
            "defaultIsSelected": false,
            "label": "Protos",
            "options": null,
          },
          "hasOptions": false,
          "parent": "2",
        },
      ],
    },
    "hasOptions": true,
    "parent": "root",
  },
  Object {
    "$$key": "3",
    "attributes": Object {
      "defaultIsSelected": false,
      "label": "Cartoon Network",
      "options": null,
    },
    "hasOptions": false,
    "parent": "root",
  },
  Object {
    "$$key": "4",
    "attributes": Object {
      "defaultIsSelected": true,
      "label": "Dragon Ball",
      "options": null,
    },
    "hasOptions": false,
    "parent": "root",
  },
]
`;

exports[`ListBoxBrowser should return a data with proper keys 2`] = `
Object {
  "current": Object {
    "0/0": Array [
      Object {
        "$$key": "0/0/1",
        "attributes": Object {
          "defaultIsSelected": true,
          "label": "The Punisher",
          "options": null,
        },
        "hasOptions": false,
        "parent": "0/0",
      },
      Object {
        "$$key": "0/0/8",
        "attributes": Object {
          "defaultIsSelected": true,
          "label": "Bucky",
          "options": null,
        },
        "hasOptions": false,
        "parent": "0/0",
      },
    ],
    "1/2": Array [
      Object {
        "$$key": "1/2/2",
        "attributes": Object {
          "defaultIsSelected": true,
          "label": "Shazam",
          "options": null,
        },
        "hasOptions": false,
        "parent": "1/2",
      },
      Object {
        "$$key": "1/2/3",
        "attributes": Object {
          "defaultIsSelected": true,
          "label": "Highfather",
          "options": null,
        },
        "hasOptions": false,
        "parent": "1/2",
      },
    ],
    "root": Array [
      Object {
        "$$key": "4",
        "attributes": Object {
          "defaultIsSelected": true,
          "label": "Dragon Ball",
          "options": null,
        },
        "hasOptions": false,
        "parent": "root",
      },
    ],
  },
}
`;

exports[`ListBoxBrowser should return option with $$key === 0 1`] = `
Object {
  "$$key": "0",
  "attributes": Object {
    "defaultIsSelected": false,
    "label": "Marvel Universe",
    "options": Array [
      Object {
        "$$key": "0/0",
        "attributes": Object {
          "defaultIsSelected": false,
          "label": "Humans",
          "options": Array [
            Object {
              "$$key": "0/0/0",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Black Cat",
                "options": Array [
                  Object {
                    "$$key": "0/0/0/0",
                    "attributes": Object {
                      "defaultIsSelected": false,
                      "label": "Akhenaten",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                  Object {
                    "$$key": "0/0/0/1",
                    "attributes": Object {
                      "defaultIsSelected": false,
                      "label": "Alistair Smythe",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                  Object {
                    "$$key": "0/0/0/2",
                    "attributes": Object {
                      "defaultIsSelected": false,
                      "label": "Angelo Fortunato",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                  Object {
                    "$$key": "0/0/0/3",
                    "attributes": Object {
                      "defaultIsSelected": false,
                      "label": "Arcade",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                  Object {
                    "$$key": "0/0/0/4",
                    "attributes": Object {
                      "defaultIsSelected": false,
                      "label": "Aria",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                  Object {
                    "$$key": "0/0/0/5",
                    "attributes": Object {
                      "defaultIsSelected": false,
                      "label": "Baal",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                ],
              },
              "hasOptions": true,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/1",
              "attributes": Object {
                "defaultIsSelected": true,
                "label": "The Punisher",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/2",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Silver Sable",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/3",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Red Skull",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/4",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Vulture",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/5",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Falcon",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/6",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Kraven the hunter",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/7",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Bruce banner",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/8",
              "attributes": Object {
                "defaultIsSelected": true,
                "label": "Bucky",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/9",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Moon Knight",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/10",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "KingPin",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/11",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Doctor Octopus",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/12",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Elektra",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/13",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Black Widow",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/14",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Hawkeye",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/15",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Bullseye",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/16",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "TaskMaster",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/17",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Star-Lord",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/18",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "DareDevil",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/19",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Ikari",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/20",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "War Machine",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/21",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Nick Fury",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/22",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Hank Pym",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/23",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Captain America",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/24",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Iron Man",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
          ],
        },
        "hasOptions": true,
        "parent": "0",
      },
      Object {
        "$$key": "0/1",
        "attributes": Object {
          "defaultIsSelected": false,
          "label": "Mutants",
          "options": null,
        },
        "hasOptions": false,
        "parent": "0",
      },
      Object {
        "$$key": "0/2",
        "attributes": Object {
          "defaultIsSelected": false,
          "label": "Celestials",
          "options": null,
        },
        "hasOptions": false,
        "parent": "0",
      },
    ],
  },
  "hasOptions": true,
  "parent": "root",
}
`;

exports[`ListBoxBrowser should return option with $$key === 1 1`] = `
Object {
  "$$key": "1",
  "attributes": Object {
    "defaultIsSelected": false,
    "label": "DC Universe",
    "options": Array [
      Object {
        "$$key": "1/0",
        "attributes": Object {
          "defaultIsSelected": false,
          "label": "Humans",
          "options": null,
        },
        "hasOptions": false,
        "parent": "1",
      },
      Object {
        "$$key": "1/1",
        "attributes": Object {
          "defaultIsSelected": false,
          "label": "Meta Humans",
          "options": null,
        },
        "hasOptions": false,
        "parent": "1",
      },
      Object {
        "$$key": "1/2",
        "attributes": Object {
          "defaultIsSelected": false,
          "label": "The Quintessence",
          "options": Array [
            Object {
              "$$key": "1/2/0",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Zeus",
                "options": null,
              },
              "hasOptions": false,
              "parent": "1/2",
            },
            Object {
              "$$key": "1/2/1",
              "attributes": Object {
                "defaultIsSelected": false,
                "label": "Ganthet",
                "options": null,
              },
              "hasOptions": false,
              "parent": "1/2",
            },
            Object {
              "$$key": "1/2/2",
              "attributes": Object {
                "defaultIsSelected": true,
                "label": "Shazam",
                "options": null,
              },
              "hasOptions": false,
              "parent": "1/2",
            },
            Object {
              "$$key": "1/2/3",
              "attributes": Object {
                "defaultIsSelected": true,
                "label": "Highfather",
                "options": null,
              },
              "hasOptions": false,
              "parent": "1/2",
            },
          ],
        },
        "hasOptions": true,
        "parent": "1",
      },
    ],
  },
  "hasOptions": true,
  "parent": "root",
}
`;

exports[`ListBoxBrowser should return option with $$key === 1/2/1 1`] = `
Object {
  "$$key": "1/2/1",
  "attributes": Object {
    "defaultIsSelected": false,
    "label": "Ganthet",
    "options": null,
  },
  "hasOptions": false,
  "parent": "1/2",
}
`;

exports[`ListBoxBrowser should return option with $$key === 2/1 1`] = `
Object {
  "$$key": "2/1",
  "attributes": Object {
    "defaultIsSelected": false,
    "label": "Zergs",
    "options": null,
  },
  "hasOptions": false,
  "parent": "2",
}
`;
